### ISSUE TITLE
Some question about training the own model

### ISSUE THREAD

**Issue by Beelotus** [PeizhiShi/SsdNet#2]
Hi,
I'm interested in your project, that's totally awesome, and I want to train the model by myself.
I follow the steps, train the model, but the performance is not well.

A few step I'm not really sure is the third step, STL models into voxel models. I run in Windows system, the parameters I used is "binvox -c -d 64 2_20.STL" (for example) for training. 
The TrSet image and label likes this:
![img](https://user-images.githubusercontent.com/92623095/137859942-1de1cd3f-604f-4dc6-9810-08c917fd72a5.png)

First, I train the 3 classes(through hole, blind hole, rectangular_passage), and change the code(data/config/'num_classes' : 4) num_classes to 4 .
I create the 2100000 TrSet and 1100 ValSet, and train the model.

The first epoch about 20000 iteration weights data(VOC) can find the poisition, but incorrect recognition :
![result1_](https://user-images.githubusercontent.com/92623095/137866687-85a5791d-4d7b-4c7d-ba38-eab4a1be722f.png)

![result2_](https://user-images.githubusercontent.com/92623095/137866719-fff14130-e1b5-48a5-a286-db380b67c05b.png)

![result3_](https://user-images.githubusercontent.com/92623095/137866734-5d9454bf-d445-4e44-b9ab-af58a116a1fb.png)

But after training for 4 epoch, the loss is maintain 1.0~2.8 can not be lower.
And the performance is worset than previous epoch.
The final result like this:
![result1](https://user-images.githubusercontent.com/92623095/137864070-e73b1505-646e-4453-8e73-c2750b6d8972.png)

![result2](https://user-images.githubusercontent.com/92623095/137864083-94a82c68-6ee6-4c61-a249-8be47905eda3.png)

![result3](https://user-images.githubusercontent.com/92623095/137864098-d744db56-a3a9-4a9b-ab3e-3fb3029f6176.png)

I wonder know what step or I didn,t notice.  
Is the training data too much for 3 classes?  STL models into voxel models is wrong?  where is the code I need to change but I didn,t notice

Best regards

**Comment by PeizhiShi** [comment]
Hi, it is great to hear that you are interested in our research paper. 

In your current setting, there are four outputs (1 background class, 3 feature classes). Those three classes in the network refer to the FIRST THREE feature classes defined in FeatureNet paper (i.e. 'o-ring', 'through hole' and 'blind hole'). In your training set, however, there are 24 feature classes. (From the given image, it can be found that there are two triangular passages, one blind hole, and one rectangular passage.) During testing, you want to recognize other features (i.e. 'through hole', 'blind hole', 'rectangular passage'), which is impossible.

In summary, the problem is that you defined a 3 class network that predicts the first three feature types, constructed a dataset that consists of 24 machining feature types for training, and aimed to recognize a new feature type (i.e. 'rectangular passage') which is not defined in the network.

Here is the solution:

num_classes in data/config.py is only utilized to control the number of outputs in a neural network. If you want to change the meaning of each output and train a 3 classes network by yourself, the following functions also need to be changed:

create_partition in /data/voc0712.py for training/validation set generation
achieve_random_model in data/voc0712.py for training/validation set generation
test.py and visualize.py for visualization
...

In those functions, you need to construct training and validation sets by using only those three types of features (i.e. 'through hole', 'blind hole', 'rectangular passage') rather than 24 types of features, and change their predefined feature ids into new ones.

My suggestion is not to change anything at the moment. Just train a 24+1 class model by using my original code, and test whether it could give you the expected result in your environment. If it works, you could adjust it based on your needs.

Best regards,
Peizhi
